File: Linear/Dynamic/Linked List/Singly Linked List/02_Inserting/End/Inserting.java
// Define a node in the linked list
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class InsertingEnd {
    // Function to insert a new node at the end of the list
    public static Node insertAtEnd(Node head, int x) {
        Node newNode = new Node(x);     // Create a new node

        /*If the Linked List is empty, make
        the new node as the head and return*/
        if (head == null)
        return newNode;

        // Store the head reference in a temporary variable
        Node last = head;

        // Traverse till the last node
        while (last.next != null) {
            last = last.next;
        }

        /*Change the next pointer of the
        last node to point to the new node*/
        last.next = newNode;

        // Return the head of the list
        return head;
    }

    public static void printList(Node head) {
        while (head != null) {
            System.out.print(head.data);
            if (head.next != null)
            System.out.print(" -> ");
            head = head.next;
        }

        System.out.println();
    }

    public static void main(String[] args) {
        // Create the linked list: 1 -> 2 -> 3 -> 4 -> 5
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);
        head.next.next.next.next = new Node(5);
        int x = 6;
        System.out.println("Linked List before insertion");
        printList(head);

        // Insert a new node at the end of the list
        head = insertAtEnd(head, x);
        
        System.out.println("Linked List after insertion");
        printList(head);
    }
}

File: Linear/Dynamic/Linked List/Circular Linked List/01_Inserting/End/Inserting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insertAtBeginning(tail, x):
    newNode = Node(x)

    if tail is None:
        newNode.next = newNode
        return newNode
    
    newNode.next = tail.next
    tail.next = newNode

    tail = newNode

    return tail

def printList(last):
    if last is None:
        return
    
    head = last.next
    temp = head

    while True:
        print(temp.data, end="")
        temp = temp.next
        if temp is not head:
            print(" -> ", end="")
        else:
            break
    
    print()

# Create circular linked list: 2 -> 3 -> 4
first = Node(2)
first.next = Node(3)
first.next.next = Node(4)

last = first.next.next
last.next = first

print("Linked List before insertion\n")
printList(last)

x = 5
last = insertAtBeginning(last, x)

print("Linked List after insertion\n")
printList(last)